service/matching/matchingEngine.go

```diff
@@ -253,11 +255,14 @@ func (e *matchingEngineImpl) updateTaskList(taskList *taskListID, mgr taskListMa
 	e.taskLists[*taskList] = mgr
 }
 
-func (e *matchingEngineImpl) removeTaskListManager(id *taskListID) {
+func (e *matchingEngineImpl) removeTaskListManager(tlMgr taskListManager) {
+	id := tlMgr.TaskListID()
 	e.taskListsLock.Lock()
-	defer e.taskListsLock.Unlock()
-
-	delete(e.taskLists, *id)
+	currentTlMgr, ok := e.taskLists[*id]
+	if ok && tlMgr == currentTlMgr {
+		delete(e.taskLists, *id)
+	}
+	e.taskListsLock.Unlock()
```

Comment:
nit: Would there be any reason to not defer this above the lock?

it's currently fine I think now, but if there's any other control flow changes It'd be easy to miss